azurecompute/src/main/java/org/jclouds/azurecompute/compute/extensions/AzureComputeSecurityGroupExtension.java

```diff
@@ -0,0 +1,460 @@
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static java.lang.String.format;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.jclouds.azurecompute.config.AzureComputeProperties.OPERATION_TIMEOUT;
+import static org.jclouds.util.Predicates2.retry;
+import java.util.List;
+import java.util.Set;
+
+import javax.annotation.Resource;
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import org.jclouds.azurecompute.AzureComputeApi;
+import org.jclouds.azurecompute.config.AzureComputeConstants;
+import org.jclouds.azurecompute.domain.CloudService;
+import org.jclouds.azurecompute.domain.Deployment;
+import org.jclouds.azurecompute.domain.NetworkConfiguration;
+import org.jclouds.azurecompute.domain.NetworkConfiguration.VirtualNetworkSite;
+import org.jclouds.azurecompute.domain.NetworkSecurityGroup;
+import org.jclouds.azurecompute.domain.Operation;
+import org.jclouds.azurecompute.domain.Role;
+import org.jclouds.azurecompute.domain.Rule;
+import org.jclouds.azurecompute.util.NetworkSecurityGroups;
+import org.jclouds.compute.domain.SecurityGroup;
+import org.jclouds.compute.domain.SecurityGroupBuilder;
+import org.jclouds.compute.extensions.SecurityGroupExtension;
+import org.jclouds.compute.reference.ComputeServiceConstants;
+import org.jclouds.domain.Location;
+import org.jclouds.logging.Logger;
+import org.jclouds.net.domain.IpPermission;
+import org.jclouds.net.domain.IpProtocol;
+
+import com.google.common.base.Function;
+import com.google.common.base.Predicate;
+import com.google.common.base.Predicates;
+import com.google.common.base.Splitter;
+import com.google.common.collect.FluentIterable;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Multimap;
+
+/**
+ * An extension to compute service to allow for the manipulation of {@link org.jclouds.compute.domain.SecurityGroup}s. Implementation
+ * is optional by providers.
+ *
+ * It considers only the custom rules added by the user and ignores the default rules created by Azure
+ */
+public class AzureComputeSecurityGroupExtension implements SecurityGroupExtension {
+
+   protected final AzureComputeApi api;
+   private final long operationTimeout;
+   private Predicate<String> operationSucceeded;
+
+   @Resource
+   @Named(ComputeServiceConstants.COMPUTE_LOGGER)
+   protected Logger logger = Logger.NULL;
+
+   @Inject
+   public AzureComputeSecurityGroupExtension(final AzureComputeApi api, @Named(OPERATION_TIMEOUT) long operationTimeout) {
+      this.api = checkNotNull(api, "api");
+      this.operationTimeout = checkNotNull(operationTimeout, "operationTimeout");
+      this.operationSucceeded = retry(new Predicate<String>() {
+         public boolean apply(String input) {
+            final Operation operation = api.getOperationApi().get(input);
+            return operation.status() == Operation.Status.SUCCEEDED;
+         }
+      }, operationTimeout, 5, 5, SECONDS);
```

Comment:
Same comment than above. Extract the predicate to a  provider and get it injected here directly.